### PR TITLE
[BEAM-2939] Fix FnApiDoFnRunner to ensure that we output within the correct window when processing a splittable dofn

### PR DESCRIPTION
**Pull request by lukecwik**
This fixes a bug where we would output within all the windows instead of just the current window.
This would not impact any SDF that used only a single window while processing.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @boyuanzz @chamikaramj

File: sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnApiDoFnRunnerTest.java [hunk 1]
```diff
@@ -1593,6 +1601,391 @@ public void testProcessElementForSizedElementAndRestriction() throws Exception {
...
+                  KV.of("7", KV.of(new OffsetRange(0, 5), GlobalWindow.TIMESTAMP_MIN_VALUE)), 2.0),
+              window1,
+              window2);
+      mainInput.accept(splitValue);
+      HandlesSplits.SplitResult trySplitResult = trySplitFuture.get();
+
+      // Check that after processing an element we don't report progress
+      assertThat(Iterables.getOnlyElement(progressRequestCallbacks).getMonitoringInfos(), empty());
+
+      // Since the SPLIT_ELEMENT is 3 we will process 0, 1, 2, 3 then be split on the first window.
```
Comment (boyuanzz) [inline sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnApiDoFnRunnerTest.java:1901]: I'm wondering why the split is on the first window?

Comment (lukecwik) [reply sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnApiDoFnRunnerTest.java:1901]: If the window is observed we need to process the element for each window independently. You do point out an issue though that the split's primary and residual roots need to account for what windows were processed and which are still remaining.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 2]
```diff
@@ -577,26 +586,83 @@ public Instant timestamp(DoFn<InputT, OutputT> doFn) {
...
-                        currentWindow,
-                        currentElement.getPane()));
+            (splitResult, watermarkEstimatorState) -> {
+              List<BoundedWindow> primaryFullyProcessedWindows =
+                  ImmutableList.copyOf(
+                      Iterables.limit(
+                          currentElement.getWindows(), currentWindowIterator.previousIndex()));
+              // Advances the iterator consuming the remaining windows.
+              List<BoundedWindow> residualUnprocessedWindows =
+                  ImmutableList.copyOf(currentWindowIterator);
```
Comment (boyuanzz) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:596]: `currentWindowIterator .next()`?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 3]
```diff
@@ -577,26 +586,83 @@ public Instant timestamp(DoFn<InputT, OutputT> doFn) {
...
       case PTransformTranslation.SPLITTABLE_PROCESS_SIZED_ELEMENTS_AND_RESTRICTIONS_URN:
         this.convertSplitResultToWindowedSplitResult =
             (splitResult, watermarkEstimatorState) -> {
+              List<BoundedWindow> primaryFullyProcessedWindows =
+                  ImmutableList.copyOf(
+                      Iterables.limit(
+                          currentElement.getWindows(), currentWindowIterator.previousIndex()));
+              // Advances the iterator consuming the remaining windows.
+              List<BoundedWindow> residualUnprocessedWindows =
+                  ImmutableList.copyOf(currentWindowIterator);
```
Comment (boyuanzz) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:641]: `currentWindowIterator.next()`?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 4]
```diff
@@ -632,6 +698,17 @@ public Object restriction() {
...
                       });
               return WindowedSplitResult.forRoots(
+                  primaryFullyProcessedWindows.isEmpty()
+                      ? null
+                      : WindowedValue.of(
+                          KV.of(
+                              KV.of(
+                                  currentElement.getValue(),
+                                  KV.of(currentRestriction, currentWatermarkEstimatorState)),
+                              fullSize),
```
Comment (boyuanzz) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:708]: I'm wondering should the size be `fullSize` * num of windows?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 5]
```diff
@@ -941,20 +1047,59 @@ private Progress getProgress() {
           convertSplitResultToWindowedSplitResult.apply(result, watermarkAndState.getValue());
     }
 
+    List<BundleApplication> primaryRoots = new ArrayList<>();
```
Comment (boyuanzz) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:1050]: This also takes care of self checkpoint, right?

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 2 above)
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:596]: ImmutableList.copyOf(iterator) drains all the remaining elements from iterator. It is effectively:
```
while (iterator.hasNext()) {
  list.add(iterator.next());
}
```

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 3 above)
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:641]: ImmutableList.copyOf(iterator) drains all the remaining elements from iterator. It is effectively:
```
while (iterator.hasNext()) {
  list.add(iterator.next());
}
```

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 4 above)
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:708]: I don't think so but this warrants a larger discussion about what does an element+restriction in multiple windows mean and how that it impacts splitting/sizing.

I would like to leave it as is until we can build that consensus.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 5 above)
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:1050]: Yes. As shown in the test the self checkpoint will checkpoint for the "remaining" windows as well.

I would like to leave it as is until we can build consensus around what we should do in these use cases.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java [hunk 6]
```diff
@@ -296,6 +299,12 @@ public void accept(WindowedValue input) throws Exception {
   /** Only valid during {@code processElement...} methods, null otherwise. */
   private WindowedValue<InputT> currentElement;
 
+  /**
+   * Only valid during {@link #processElementForSizedElementAndRestriction} and {@link
+   * #processElementForSizedElementAndRestriction}.
```
[the remark this answers is not in the document]
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:304]: Fixed.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java (hunk 2 above)
Comment (boyuanzz) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:596]: Ah I see. It's different from what I know about an `Iterator`. Thanks!

**Comment by lukecwik** [comment]
Run Java PreCommit

**Pull request merged**